### PR TITLE
New Track Seed Objects

### PR DESCRIPTION
**Pull request by osbornjd**
This is a major change to the track reconstruction workflow introducing several new `TrackSeed` DST objects into the reconstruction chain. The `TrackSeed` objects are significantly more lightweight than the corresponding `SvtxTrack` objects and have corresponding member variables better suited to our seeding algorithms. There are a number of notable changes:

1. `TrackSeed` objects can circle fit themselves, so we don't duplicate circle fitting code
2. The final creation from the seeding chain is a `SvtxTrackSeed` which contains two indices corresponding to the relevant TPC and silicon seed. Any TPOT matches are included in the TPC seed.
3. The containers that store `TrackSeeds` are vectors. The `TrackSeedContainer` object has utility functions that allow one to look up the entry for a particular `TrackSeed` based on it's corresponding iterator. Therefore I recommend NOT iterating through the track map with range based iterators as it doesn't give you as clean of access to the entry. As in:
```
for(const auto& track : *m_seedMap) // Don't do
{
   // Do stuff
}
for(auto iter = m_seedMap->begin(); iter != m_seedMap->end(); ++iter) // should do
{
    TrackSeed *track = *iter;
    unsigned int seedID = m_seedMap->index(iter);
    // Do stuff
}

```
4. `PHGhostRejection` is now a utility module that is called directly from the ALICE KF Propagator

This is a major modification to the track reconstruction and SHOULD NOT be merged until I (and Jenkins) give the okay to do so.

## Types of changes
[comment]: <> ( What types of changes does your code introduce? Put an `x` in all the boxes that apply: )
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work for users)
- [x] Requiring change in macros repository (Please provide links to the macros pull request in the last section)
- [x] I am a member of [GitHub organization of sPHENIX Collaboration](https://github.com/orgs/sPHENIX-Collaboration/people), EIC, or ECCE (contact Chris Pinkenburg to join)

## What kind of change does this PR introduce? (Bug fix, feature, ...)

[comment]: <> ( What does this PR do? Linking to talk in software meeting encouraged )

## TODOs (if applicable)

Further changes will be to introduce a new `SvtxTrack` object which significantly cuts down the associated member variables for final tracks.

### CODE REVIEW

File: offline/packages/trackbase_historic/SvtxTrackMap_v2.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  bool empty() const override{ return _map.empty(); }
+  size_t size() const override { return _map.size(); }
+  size_t count(unsigned int idkey) const override{ return _map.count(idkey); }
+  void clear() override { Reset(); }
+
+  const SvtxTrack* get(unsigned int idkey) const override;
+  SvtxTrack* get(unsigned int idkey) override;
+  SvtxTrack* insert(const SvtxTrack* track) override;
+  SvtxTrack* insert(const SvtxTrack* track, unsigned int index) override;
+  size_t erase(unsigned int idkey) override
```
Comment (hupereir) [inline offline/packages/trackbase_historic/SvtxTrackMap_v2.h:34]: Same remark: shouldn't you also delete the actual track ?

Comment (osbornjd) [reply offline/packages/trackbase_historic/SvtxTrackMap_v2.h:34]: This is an exact copy of the current `SvtxTrackMap_v1.h` object, so if it isn't a problem there it won't be here either. Did we have leaks in the current object? Besides, isn't it deleted with the delete call or what am I missing?

File: offline/packages/trackbase_historic/TrackSeedContainer_v1.h [hunk 2]
```diff
@@ -28,19 +28,23 @@ class TrackSeedContainer_v1 : public TrackSeedContainer
   const TrackSeed* get(const std::size_t key) const override;
   TrackSeed* get(const std::size_t key) override;
   TrackSeed* insert(const TrackSeed* seed) override;
-  Iter erase(const std::size_t key) override
+  void erase(const std::size_t key) override
     {
       delete m_seeds.at(key);
-      Iter iter = m_seeds.begin() + key;
-      return m_seeds.erase(iter);
+      m_seeds.at(key) = nullptr;
```
[the remark this answers is not in the document]
Comment (osbornjd) [reply offline/packages/trackbase_historic/TrackSeedContainer_v1.h:34]: Isn't that what `delete m_seeds.at(key)` does, or are you referring to something else?

**Comment by osbornjd** [comment]
Any comments on this PR are welcome, there are a lot of changes and we should get it as right as we can before merging it.

**Comment by pinkenburg** [comment]
tiny merging conflict in offline/packages/trackreco/PHSimpleKFProp.h

**Comment by pinkenburg** [comment]
> @blackcathj or @pinkenburg can you explain the Jenkins clang errors to me? I didn't change these files and doing a grep in my local branch does not yield any instances of `ActsSurfaceMaps` or `ActsTrackingGeometry` being declared as a class

I had a look - this warning gets emitted if there is also a definition of ActsSurfaceMaps as a class. Indeed a 
find . -type f -exec grep ActsSurfaceMaps {} \; -ls
gives quite a few hits one of them 
./simulation/g4simulation/g4eval/TrackSeedTrackMapConverter.h
declares this as class
class ActsSurfaceMaps;
  ActsSurfaceMaps *m_surfmaps = nullptr;

I assume for ActsTrackingGeometry it is similar. It seems to be valid code (except for microsoft) but having two definitions for the same thing or the same name for two different things is just not good

**Comment by osbornjd** [comment]
This now provides similar timing performance to the current main branch, so I am marking it ready to merge assuming Jenkins agrees

**Comment by osbornjd** [comment]
The cppcheck new warning seems to be a warning that was already there for the first implementation of the SvtxTrackMap_v1 but was looked over [here](https://web.sdcc.bnl.gov/jenkins-sphenix/job/sPHENIX/job/cpp-check-pipeline/3476//cppcheck/folder.-1764622961/fileName.-1316341844/source.dc570738-dc7e-41a5-9cdc-75fc38e00cfa/#47). I've found several instances online of this being reported as a bug in several versions of cppcheck - what version are we running? See e.g. [here](https://sourceforge.net/p/cppcheck/discussion/general/thread/b18f7aaf/?limit=25)

**Comment by pinkenburg** [comment]
macro dies with
 circle/circle intersection calculation failed, skip this cluster
 target_radius 55.625 fitted R 12.1182 fitted X0 37.3576 fitted Y0 21.2524
 circle/circle intersection calculation failed, skip this cluster
 target_radius 56.875 fitted R 12.1182 fitted X0 37.3576 fitted Y0 21.2524
/var/lib/jenkins/workspace/sPHENIX/Build-Master-gcc8/coresoftware/offline/framework/fun4all/Fun4AllServer.cc:603:  caught exception thrown during process_event from PHSimpleKFProp
error: vector::_M_range_check: __n (which is 1) >= this->size() (which is 1)

**Comment by osbornjd** [comment]
I'll have to try and reproduce the exact macro error, I just tested and things run for me locally...

**Pull request merged**